akka-stream-tests/src/test/scala/akka/stream/io/StreamTcpSpec.scala

```diff
@@ -77,36 +79,142 @@ class StreamTcpSpec extends AkkaSpec with TcpHelper {
       Source(tcpWriteProbe.publisherProbe).via(StreamTcp().outgoingConnection(server.address).flow).to(Sink(tcpReadProbe.subscriberProbe)).run()
       val serverConnection = server.waitAccept()
 
+      // Client can still write
+      tcpWriteProbe.write(testData)
+      serverConnection.read(5)
+      serverConnection.waitRead() should be(testData)
+
+      // Close client side write
       tcpWriteProbe.close()
-      // FIXME: expect PeerClosed on server
+      serverConnection.expectClosed(PeerClosed)
+
+      // Server can still write
       serverConnection.write(testData)
       tcpReadProbe.read(5) should be(testData)
+
+      // Close server side write
       serverConnection.confirmedClose()
       tcpReadProbe.subscriberProbe.expectComplete()
     }
 
-    "stop reading when the input stream is cancelled" in {
+    "work when remote closes write, then client closes write" in {
+      val testData = ByteString(1, 2, 3, 4, 5)
       val server = new Server()
+
+      val tcpWriteProbe = new TcpWriteProbe()
+      val tcpReadProbe = new TcpReadProbe()
+      Source(tcpWriteProbe.publisherProbe).via(StreamTcp().outgoingConnection(server.address).flow).to(Sink(tcpReadProbe.subscriberProbe)).run()
+      val serverConnection = server.waitAccept()
+
+      // Server can still write
+      serverConnection.write(testData)
+      tcpReadProbe.read(5) should be(testData)
+
+      // Close server side write
+      serverConnection.confirmedClose()
+      tcpReadProbe.subscriberProbe.expectComplete()
+
+      // Client can still write
+      tcpWriteProbe.write(testData)
+      serverConnection.read(5)
+      serverConnection.waitRead() should be(testData)
+
+      // Close client side write
+      tcpWriteProbe.close()
+      serverConnection.expectClosed(ConfirmedClosed)
+    }
+
+    "work when client closes read, then client closes write" in {
       val testData = ByteString(1, 2, 3, 4, 5)
+      val server = new Server()
 
       val tcpWriteProbe = new TcpWriteProbe()
       val tcpReadProbe = new TcpReadProbe()
       Source(tcpWriteProbe.publisherProbe).via(StreamTcp().outgoingConnection(server.address).flow).to(Sink(tcpReadProbe.subscriberProbe)).run()
       val serverConnection = server.waitAccept()
 
-      tcpReadProbe.close()
-      // FIXME: expect PeerClosed on server
+      // Server can still write
       serverConnection.write(testData)
-      tcpReadProbe.subscriberProbe.expectNoMsg(1.second)
+      tcpReadProbe.read(5) should be(testData)
+
+      // Close client side read
+      tcpReadProbe.tcpReadSubscription.cancel()
+
+      // Client can still write
+      tcpWriteProbe.write(testData)
       serverConnection.read(5)
+      serverConnection.waitRead() should be(testData)
+
+      // Close client side write
+      tcpWriteProbe.close()
+      // FIXME: Why not Closed?
+      serverConnection.expectClosed(PeerClosed)
```

Comment:
I think this is a bug in IO, after a PeerClosed a full close should result in Closed, not another PeerClosed